components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java

```diff
@@ -1277,31 +1289,41 @@ private void updateArtifactCount(String businessRuleUUID, int artifactCount)
-            String instanceLimit = getRuleTemplate(businessRuleFromTemplate).getInstanceCount();
-            count = getInstanceCount(businessRuleFromTemplate.getRuleTemplateUUID());
-            if ("one".equalsIgnoreCase(instanceLimit) && count >= 1) {
-                throw new TemplateInstanceCountViolationException("Rule Template '" +
-                        ((BusinessRuleFromTemplate) businessRule).getRuleTemplateUUID() + "' can be instantiated " +
-                        "only once and have been already instantiated.");
-            }
-        } else if (businessRule instanceof BusinessRuleFromScratch) {
-            BusinessRuleFromScratch businessRuleFromScratch = (BusinessRuleFromScratch) businessRule;
-            List<RuleTemplate> ruleTemplates =
-                    (List<RuleTemplate>) getInputOutputRuleTemplates(businessRuleFromScratch);
-            String instanceLimit;
-            for (RuleTemplate ruleTemplate : ruleTemplates) {
-                instanceLimit = (ruleTemplate.getInstanceCount());
-                count = getInstanceCount(ruleTemplate.getUuid());
+        int count;
+        int deploymentStatus = 0;
+        try {
+            deploymentStatus = queryExecutor.executeRetrieveDeploymentStatus(businessRule.getUuid());
+        } catch (BusinessRulesDatasourceException e) {
+            throw new TemplateManagerServiceException("Failed to retrieve the status of the business rule '" +
+                    businessRule.getUuid() + "'");
+        }
+        if ((deploymentStatus != TemplateManagerConstants.DEPLOYED) &&
+                (deploymentStatus != TemplateManagerConstants.PARTIALLY_DEPLOYED)) {
+            if (businessRule instanceof BusinessRuleFromTemplate) {
+                BusinessRuleFromTemplate businessRuleFromTemplate = (BusinessRuleFromTemplate) businessRule;
+                String instanceLimit = getRuleTemplate(businessRuleFromTemplate).getInstanceCount();
+                count = getInstanceCount(businessRuleFromTemplate.getRuleTemplateUUID());
                 if ("one".equalsIgnoreCase(instanceLimit) && count >= 1) {
                     throw new TemplateInstanceCountViolationException("Rule Template '" +
-                            (ruleTemplate.getUuid() + "' can be instantiated " +
-                                    "only once and have been already instantiated."));
+                            ((BusinessRuleFromTemplate) businessRule).getRuleTemplateUUID() + "' can be instantiated " +
+                            "only once and have been already instantiated.");
+                }
+            } else if (businessRule instanceof BusinessRuleFromScratch) {
+                BusinessRuleFromScratch businessRuleFromScratch = (BusinessRuleFromScratch) businessRule;
+                List<RuleTemplate> ruleTemplates =
+                        (List<RuleTemplate>) getInputOutputRuleTemplates(businessRuleFromScratch);
```

Comment:
Can we change this method to return a List?

Reply:
Fixed with PR #1205